src/rust/engine/options/src/env.rs

```diff
@@ -14,13 +15,46 @@ pub struct Env {
   pub(crate) env: HashMap<String, String>,
 }
 
+#[derive(Debug)]
+pub struct DroppedEnvVars {
+  pub non_utf8_keys: Vec<String>,
+  pub keys_with_non_utf8_values: Vec<String>,
+}
+
 impl Env {
   pub fn new(env: HashMap<String, String>) -> Self {
     Self { env }
   }
 
-  pub fn capture() -> Self {
-    Self::new(env::vars().collect::<HashMap<_, _>>())
+  pub fn capture_lossy() -> (Self, DroppedEnvVars) {
+    Self::do_capture_lossy(env::vars_os())
+  }
+
+  pub(crate) fn do_capture_lossy<I>(env_os: I) -> (Self, DroppedEnvVars)
+  where
+    I: Iterator<Item = (OsString, OsString)>,
+  {
+    let mut env: HashMap<String, String> = HashMap::with_capacity(env_os.size_hint().0);
+    let mut dropped = DroppedEnvVars {
+      non_utf8_keys: Vec::new(),
+      keys_with_non_utf8_values: Vec::new(),
+    };
+    for (os_key, os_val) in env_os {
+      match (os_key.into_string(), os_val.into_string()) {
+        (Ok(key), Ok(val)) => {
+          env.insert(key, val);
+        }
+        (Ok(key), Err(_)) => dropped.keys_with_non_utf8_values.push(key),
+        // We'll only be able to log the lossy name of any non-UTF-8 keys, but
```

Comment:
This is a non-local comment - no logging here. I think it'd be better to just store `OsString` in `DroppedEnvVars` and let the user decide what to do in the context of their use. They can convert to String and sort if they so desire or just print OsString debug content, or - whatever. They have everything they need.